### ISSUE TITLE
Make matplotlib work

### ISSUE THREAD

**Issue by kennethcassel** [slipHQ/run-wasm#24]
Make a simple matplotlib script work in example-nextjs

Should be able to get a simple example going https://matplotlib.org/stable/gallery/lines_bars_and_markers/lines_with_ticks_demo.html#sphx-glr-gallery-lines-bars-and-markers-lines-with-ticks-demo-py

I think we might need to use something besides the monaco code editor to display this code.

```python
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import patheffects

fig, ax = plt.subplots(figsize=(6, 6))
ax.plot([0, 1], [0, 1], label="Line",
        path_effects=[patheffects.withTickedStroke(spacing=7, angle=135)])

nx = 101
x = np.linspace(0.0, 1.0, nx)
y = 0.3*np.sin(x*8) + 0.4
ax.plot(x, y, label="Curve", path_effects=[patheffects.withTickedStroke()])

ax.legend()

plt.show()```

**Comment by mcintyre94** [comment]
I was going to see if you were interested in getting this upstreamed! 

I got this working by loading this Python code:

```python
import matplotlib.pyplot as plt

f = plt.figure()

def get_render_element(self):
    return document.getElementById('canvas-figure')

f.canvas.create_root_element = get_render_element.__get__(
    get_render_element, f.canvas.__class__
)
```

And adding a `<div id="canvas-figure" />` as output

Reference is this SO comment: https://stackoverflow.com/a/59571016/1375972

Caveats I haven't fixed yet:

- I get an error about it trying to add a context menu to a div that is `None`. Might just need to give it some more structure rather than an empty div, but might need more digging
- It requires font-awesome icons, I didn't get around to figuring out how to make them work yet

Some other things to consider:

- Would we want the flexibility to show multiple plots? The SO gives an alternative `get_render_element` implementation that appends (to body, but we could presumably append to a div):

```python
def create_root_element1(self):
    div = document.createElement('div')
    document.body.appendChild(div)
    return div
```

- We probably want to make this opt-in because importing matplotlib is expensive
- Not sure whether the font-awesome dependency belongs in run-wasm or example-nextjs

This is where I got to earlier using the above code: https://twitter.com/callum_codes/status/1440594576922714117?s=20

**Comment by kennethcassel** [comment]
Great info @mcintyre94! 

A more meta question is:

What other languages or situations will require output in a canvas or similar element? 

Is this something we want to handle at the language level or is there a higher level approach we can take which would solve this for, say Python, Ruby, typescript, etc

**Comment by hatemhosny** [comment]
Hello, 
Great project 👍 

This is how I got matplotlib to work with Pyodide 

https://dev.livecodes.io/?template=pyodide

(this is the unreleased development build of [LiveCodes](https://github.com/live-codes/livecodes))

The code sample does make use of `figure.canvas.create_root_element` and adds font-awesome icons.

I hope this helps.

**Comment by mcintyre94** [comment]
@hatemhosny Thanks for sharing, that looks great! LiveCodes looks really interesting too 👀 

@kennethcassel that's definitely a good point. I noticed the AssemblyScript you linked in #13 has some cool canvas demos so I'm sure there'd be a similar issue there too around integration. Probably worth putting this on hold until we have a few languages working and can try to figure out if there's something sensible to share?

In the mean time though I'm going to get it working for my site with whatever hackery I need, because it's really cool! I'm happy to add a separate page to `example-nextjs` once I have it working to at least document how to do it for now, if that works for you?

**Comment by kennethcassel** [comment]
Great work on this @mcintyre94! 

Closed with PR #41
